File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/scope/Scope.scala
package io.github.chaosunity.casc.parsing.scope

import com.google.common.collect.Lists
import io.github.chaosunity.casc.exceptions.{LocalVariableNotFoundException, MethodSignatureNotFoundException}
import io.github.chaosunity.casc.parsing.`type`.{BuiltInType, ClassType, Type}
import io.github.chaosunity.casc.parsing.expression.Expression
import io.github.chaosunity.casc.parsing.global.Metadata
import io.github.chaosunity.casc.util.ReflectionMapper
import org.apache.commons.lang3.reflect.{ConstructorUtils, MethodUtils}
import org.apache.commons.lang3.{ClassUtils, RegExUtils}

import java.util
import java.util.{Collections, Optional}
import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters.{CollectionHasAsScala, SeqHasAsJava}

class Scope(private val _metadata: Metadata) {
    private var _functionSignatures: ListBuffer[FunctionSignature] = ListBuffer()
    private var _localVariables: ListBuffer[LocalVariable] = ListBuffer()

    def this(scope: Scope) = {
        this(scope._metadata)
        _functionSignatures = scope._functionSignatures.clone()
        _localVariables = scope._localVariables.clone()
    }

    def addSignature(signature: FunctionSignature): Unit =
        _functionSignatures += signature

    def parameterLessSignatureExists(identifier: String): Boolean =
        signatureExists(identifier, Collections.emptyList())

    def signatureExists(identifier: String, parameters: util.List[Type]): Boolean = {
        if (identifier.equals("super")) return true
        _functionSignatures.forall(_.matches(identifier, parameters))
    }

    def getMethodCallSignatureWithoutParameters(identifier: String): FunctionSignature =
        getMethodCallSignature(identifier, new util.ArrayList[Type]())

    def getMethodCallSignature(identifier: String, arguments: util.Collection[Expression]): FunctionSignature =
        getMethodCallSignature(identifier, arguments.asScala.map(_.`type`).toList.asJava)

    def getMethodCallSignature(identifier: String, parameterTypes: util.List[Type]): FunctionSignature =
        if (identifier.equals("super")) new FunctionSignature(identifier, Lists.newArrayList(), BuiltInType.VOID)
        else _functionSignatures.find(_.matches(identifier, parameterTypes)).getOrElse(throw new MethodSignatureNotFoundException(this, identifier, parameterTypes))

    def getSignatureOnClassPath(fullMethodName: String): Optional[FunctionSignature] = {
        val methodName = RegExUtils.removePattern(fullMethodName, ".*\\.")
        val className = fullMethodName

        var methodOwnerClass: Class[_] = null

        try {
            methodOwnerClass = ClassUtils.getClass(className)
        } catch {
            case e: ClassNotFoundException => throw e
        }

        val accessibleMethod = MethodUtils.getAccessibleMethod(methodOwnerClass, methodName)

        if (accessibleMethod != null) {
            val signature = ReflectionMapper.fromMethod(accessibleMethod)

            return Optional.of(signature)
        }

        val accessibleConstructor = ConstructorUtils.getAccessibleConstructor(methodOwnerClass)

        if (accessibleConstructor != null) {
            val signature = ReflectionMapper.fromConstructor(accessibleConstructor)

            return Optional.of(signature)
        }

        Optional.empty()
    }


    def addLocalVariable(localVariable: LocalVariable): Unit =
        _localVariables += localVariable

    def getLocalVariable(variableName: String): LocalVariable =
        _localVariables.find(_.name.equals(variableName))
            .getOrElse(throw new LocalVariableNotFoundException(this, variableName))

    def localVariableExists(variableName: String): Boolean =
        _localVariables.forall(_.name.equals(variableName))

    def getLocalVariableIndex(variableName: String): Int =
        _localVariables.indexOf(getLocalVariable(variableName))

    def className: String = _metadata.className

    def classInternalName: String = classType.internalName

    def superClassInternalName: String = new ClassType(superClassName).internalName

    def classType: Type = new ClassType(className)

    private def superClassName: String = _metadata.superClassName
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/scope/Scope.kt
package io.github.chaosunity.casc.parsing.scope

import io.github.chaosunity.casc.parsing.MetaData
import io.github.chaosunity.casc.parsing.node.expression.Argument
import io.github.chaosunity.casc.parsing.node.expression.SuperCall
import io.github.chaosunity.casc.parsing.type.BuiltInType
import io.github.chaosunity.casc.parsing.type.ClassType
import io.github.chaosunity.casc.parsing.type.Type

class Scope(private val metadata: MetaData) {
    val localVariables = linkedMapOf<String, LocalVariable>()
    val fields = linkedMapOf<String, Field>()
    val functionSignatures = mutableListOf<FunctionSignature>()

    var callingScope = CallingScope.STATIC

    val className = metadata.className
    val classType = ClassType(className)
    val superClassInternalName = ClassType(metadata.superClassName).internalName

    constructor(scope: Scope) : this(scope.metadata) {
        callingScope = scope.callingScope
        localVariables += scope.localVariables
        fields += scope.fields
        functionSignatures += scope.functionSignatures
    }

    fun addSignature(signature: FunctionSignature) {
        if (isParameterLessSignatureExists(signature.name)) throw RuntimeException("Function '${signature.name}' already exists.")
        functionSignatures += signature
    }

    fun isParameterLessSignatureExists(identifier: String): Boolean =
        isSignatureExists(identifier, listOf())

    fun isSignatureExists(identifier: String, arguments: List<Argument>): Boolean =
        if (identifier == SuperCall.SUPER_CALL_ID) true
        else functionSignatures.any { it.matches(identifier, arguments) }

    fun getMethodCallSignatureWithoutParameters(identifier: String): FunctionSignature =
        getMethodCallSignature(identifier, listOf())

    fun getMethodCallSignature(identifier: String, arguments: List<Argument>): FunctionSignature =
        if (identifier == SuperCall.SUPER_CALL_ID) FunctionSignature(
            SuperCall.SUPER_CALL_ID,
            listOf(),
            BuiltInType.VOID
        )
        else functionSignatures.find { it.matches(identifier, arguments) }
            ?: throw RuntimeException("Function '$identifier' does not exist.")

    fun getConstructorCallSignature(className: String, arguments: List<Argument>): FunctionSignature {
        if (className != this.className) {
            val argumentsType = arguments.map(Argument::type)

            return ClassPathScope().getConstructorSignature(className, argumentsType)
                ?: throw RuntimeException(
                    "Class constructor '$className' with type arguments '${
                        argumentsType.map(Type::internalName).joinToString(", ")
                    }' does not exist."
                )
        }

        return getMethodCallSignature(null, className, arguments)
    }

    fun getMethodCallSignature(owner: Type?, methodName: String, arguments: List<Argument>): FunctionSignature {
        if (owner != null && owner != classType) {
            val argumentsType = arguments.map(Argument::type)

            return ClassPathScope().getMethodSignature(owner, methodName, argumentsType)
                ?: throw RuntimeException("Class '$className' with type arguments '${argumentsType.joinToString(", ")}' does not exist.")
        }

        return getMethodCallSignature(methodName, arguments)
    }

    fun addField(field: Field) {
        fields += field.name to field
    }

    fun getField(fieldName: String): Field =
        fields[fieldName] ?: throw RuntimeException("Field '$fieldName' does not exist.")

    fun isFieldExists(fieldName: String): Boolean =
        fields.containsKey(fieldName)

    fun addLocalVariable(variable: LocalVariable) {
        localVariables += variable.name to variable
    }

    fun getLocalVariable(variableName: String): LocalVariable =
        localVariables[variableName] ?: throw RuntimeException("Variable '$variableName' does not exist.")

    fun isLocalVariableExists(variableName: String): Boolean =
        localVariables.containsKey(variableName)

    fun getLocalVariableIndex(variableName: String): Int =
        localVariables.keys.indexOf(variableName)
}